### ISSUE TITLE
Xenforo support in Authme 5.0 ?

### ISSUE THREAD

**Issue by Slaand** [AuthMe/AuthMeReloaded#137]
I remember, what older authme versions having xfsha256 encryption, but xenforo have separate tables: user table and password table
Is it real to return back xenforo support? Its would be nice
p.s. sorry for my english

**Comment by sgdc3** [comment]
The problem with xenforo is that it isn't open source, so we have difficulties to study his database structure.

**Comment by Slaand** [comment]
I can give clean xenforo forum with db

**Comment by sgdc3** [comment]
bump

**Comment by ljacqu** [comment]
FYI: Admin password in the dump is `oIYPsCpn` (as mentioned by @Xephi).

The code that we have for Xenforo (doesn't work!) has been pushed to the **137-xenforo-support** branch (aed23cb). There are various code snippets with `TODO #137` scattered all around the codebase that need to be tackled. In particular:
- [ ] Make sure that the salt and the hash are always retrieved and stored together
- [ ] Test AuthMe with the XENFORO hash and make sure it works (register a player, unregister a player, change a player's password...)
- [ ] Create a test for the Xenforo algorithm class. It already exists and a comment in the class explains how to create the test class (It's as simple as running a method to generate the entire class)
- [ ] Make sure there are no more `TODO #137` in the code.

**Comment by Slaand** [comment]
Also, do you know, what XF use 2 separate tables?
xf_user - user table
xf_user_authenticate - password table

And yes, for now it's doesn't work at all.
[Server thread/WARN]: [AuthMe] Unknown Hash Algorithm; defaulting to SHA256

Waiting for new builds to test them :dancer:

**Comment by ljacqu** [comment]
Thanks for the info. @DNx5 is working on adding Xenforo support. Right now it's not in AuthMe just yet, so it doesn't know the "XENFORO" hash algorithm for now. ^^

@Slaand, thanks a lot for all of your help (the information and the dump you have provided) and your patience!

**Comment by ljacqu** [comment]
@DNx5 Any news on this? I recently generated the test class for Xenforo and I think the only problem was a null pointer exception from ConsoleLogger. That's just because of the test — I'll fix this (i.e. I'll do the test class) so it doesn't stop progress on this issue

**Comment by DNx5** [comment]
I renamed XENFORO hash algorithm to XFBCRYPT because current version of xenforo is using Bcrypt but the old version is sha256/md5

**Comment by ljacqu** [comment]
That sounds nice! Thanks for all the recent commits. I'll fix the tests and assign this to you for merging to master? I don't know if there's anything else pending for Xenforo support

**Comment by ljacqu** [comment]
Tests fixed.

**Comment by sgdc3** [comment]
Ok, @Slaand can you try it? ;)

**Comment by Slaand** [comment]
1. Problem with player register, after /reg command server server snow me an error at console:
[Craft Scheduler Thread - 1/WARN]: [AuthMe] Field 'email' doesn't have a default value
So what's wrong? :D My cfg http://pastebin.com/xTuxfd4Z

2. After password change at server, player can't log in anymore. At server/website same error: wrong password.

**Comment by Slaand** [comment]
Also if you still need alot more info from this, you can take all my logins and passes from xenforo db, but admin account is broken for now, because of this passchange at server :D
P.S. Sorry for my english

**Comment by DNx5** [comment]
try to set encryption to `xfbcrypt`

**Comment by DNx5** [comment]
probably related to #418 
but I dont know why.

**Comment by Slaand** [comment]
My encryption already is XFBCRYPT, you can see my config, comments above.

**Comment by DNx5** [comment]
You can't login only after change your password right?

**Comment by Slaand** [comment]
Yes, that's right. Later I'll do more tests with xenforo and authme, maybe we get more bugs :)

**Comment by DNx5** [comment]
Is your previous password generated by AuthMe with `xfbcrypt` encryption?
Or by Xenforo?

**Comment by Slaand** [comment]
By xenforo

**Comment by DNx5** [comment]
@Slaand Should works fine now :smile:

**Comment by DNx5** [comment]
try latest dev build,
http://ci.xephi.fr/job/AuthMeReloaded/

**Comment by Slaand** [comment]
Almost perfect. 1 problem left - when player register at server, he can't log in website.
![0000](https://cloud.githubusercontent.com/assets/1496072/12213029/404f9a5e-b67a-11e5-819d-1ea50b2d2f99.PNG)

**Comment by sgdc3** [comment]
@DNx5

**Comment by DNx5** [comment]
@Slaand I'm sure that's optional.
could you check xf_user table?

**Comment by Slaand** [comment]
@DNx5 look for PM in gitter, I send info to xenforo db

**Comment by Slaand** [comment]
Oww.. I need also test out password change from website.

**Comment by sgdc3** [comment]
@ljacqu Can you update the checklist above? ;)

**Comment by sgdc3** [comment]
Can we close the issue or there is still work to do with it?

**Comment by ljacqu** [comment]
@sgdc3: Thanks for the query. @DNx5 knows best what the status is of this issue since he's implemented Xenforo support. I've updated the checklist with what I know is done. I don't know if tests have been done (though I would assume so) and I know that the readme hasn't been updated (though it doesn't _need_ to be if we don't care/want).

**Comment by sgdc3** [comment]
Added the support on the readme..

**Comment by Slaand** [comment]
Hello guys, how it's going?

**Comment by DNx5** [comment]
I thought everything is working fine, by reading your comment.

**Comment by Slaand** [comment]
No, my last comment is password change on forum, and its working.
But when players register at server, they can't login at website. So, at table xf_user_authenticate plugin does not save remember_key (screen above), maybe its an issue.

**Comment by Slaand** [comment]
[xenforo-cp.zip](https://github.com/Xephi/AuthMeReloaded/files/88827/xenforo-cp.zip)

**Comment by sgdc3** [comment]
@DNx5 Can you look at this? ;)

**Comment by sgdc3** [comment]
@Slaand @DNx5 Bump, any news?

**Comment by Slaand** [comment]
@sgdc3 hello, nothing new. But Im still available for bug testing.

**Comment by sgdc3** [comment]
@ljacqu We should fix this ASAP xD

**Comment by sgdc3** [comment]
@ljacqu Would you like to take a quick look?

**Comment by CandyManN** [comment]
Any news so far? I'm really waiting for the Xenforo Integration ^^

**Comment by sgdc3** [comment]
@ljacqu could you give this a try? It's a very requested feature

**Comment by ljacqu** [comment]
I think this is almost done, what still doesn't work is that a user cannot log into the website when he registers via AuthMe [[x](https://github.com/Xephi/AuthMeReloaded/issues/137#issuecomment-170170073)].

If everything else works, we need to know what needs to be written into the database additionally. Unfortunately, as long as no one can tell us what it is and we don't have access to the Xenforo source, we're blind.

**Comment by CandyManN** [comment]
@ljacqu I may help you to access a copy of the Xenforo source, but can we talk privately?

Thank you ^^

**Comment by Slaand** [comment]
Hello, how it's going?

**Comment by ljacqu** [comment]
Waiting for feedback from @gabriel3l :)

**Comment by gabriel3l** [comment]
Sorry, I went on vacation and I'll be back tomorrow.
As soon as I get my hands back on my computer I'll recheck the code ^^

**Comment by Eufranio** [comment]
By now, is there a way to login in minecraft with the password from Xenforo? And, is there a way to block registering in-game, to players register only in the forum? 
Also, would be cool a (WIP) wiki of that interaction, I'm very interested in the feature xD

**Comment by DNx5** [comment]
@Eufranio the current problem is, you can't login to forum when users registered in game. so, in your case it should works fine.

**Comment by Eufranio** [comment]
And where is the tutorial/wiki/how-to?

**Comment by Slaand** [comment]
I still can help with fresh copy of xenforo with db access + multicraft panel with authme plugin to test out all of this. If needed, write me pm maybe :D

**Comment by UnioDex** [comment]
Hello, i have some questions i hope you can help me:

- I want to start use XenForo with AuthMe. Right now i have 920k users with AuthMe and i have to impot all my users to XenForo database. Can you help me about that or should i hire someone to make this?
- Right now XenForo and AuthMe have problems. But if i disable in game /register, /changepassword and all AuthMe commands can i use XenForo and AuthMe together? People will register from XenForo and will make all email, password changes from there.
- Can i found a list about known Xenforo and Authme connection problems? 
- Data's is not only in one database how it is not being problem for AuthMe?

I asked so much but it is a production server not development. For this reason i have to be careful. 

Thanks.

**Comment by UnioDex** [comment]
Status Update:
- I hired someone to make import script. And import successfully worked. 
- When you disable in game AuthMe commands except /login you can use both XenForo and AuthMe properly.
- AuthMe adds password, ip, x, y, z, lastlogin fields in xf_user table. Password have to be removed. Also lastlogin not necessary. These fields dont have default value and cant be empty. So it is starting an issue. You cant register from xenforo if you dont change these options.

**Comment by sgdc3** [comment]
Duplicate of #957
